### ISSUE TITLE
Eloquent collection data available in initial render not available in livewire message nor JavaScript using livewire data property

### ISSUE THREAD

**Issue by KenLewis** [livewire/livewire#1641]
### Description
Eloquent collection data that is loaded in `mount` to a public property is able to be used to render the view but isn't available in livewire messages and are therefore not available in JavaScript using livewire data property and therefore can't be "entangled" with Alpine.

From the livewire documentation:

> Properties can ONLY be either JavaScript-friendly data types (string, int, array, boolean), OR one of the following PHP types: Collection, DateTime, Model, EloquentCollection.

While debugging this issue, I noticed 2 things:

1. While the data isn't available in `$0.__livewire.serverMemo.data` (thus can't be "entangled" with Alpine), it is available in `$0.__livewire.serverMemo.dataMeta.models`

2. If I cast the Eloquent collection to an array by using `->toArray()` in `mount`, the data is available in both the view and JavaScript. While this seems like an acceptable work around, the documentation says that an EloquentCollection should work.

### Exact steps to reproduce
- Create a public property in the livewire component (`public myProperty;`)
- In `mount`, set the property's value to Eloquent collection results (`$this->myProperty = MyModel::all();`)
- Attempt to access the data by using `Livewire.find('<COMPONENT ID>').myProperty` or by selecting the component in DevTools and using `$0.__livewire.data.myProperty` (the property exists but is an empty array)

### Stripped-down, copy-pastable code snippets
I added the `$this->agreementCollection` to show it's not related to me querying the `elements` relationship of `$agreement`

```
class Editor extends Component
{
    public $agreement;
    public $agreementCollection;
    public $elementsArray;
    public $elementsCollection;

    protected $rules = [
        'agreement.name' => 'required|string|min:6',
        'agreement.description' => 'required|string|max:500',
    ];

    public function mount(Agreement $agreement)
    {
        $this->agreement = $agreement; // App\Models\Agreement (available in JS)

        $this->elementsArray = $agreement->elements->toArray();  // Array (available in JS)

        $this->elementsCollection = $agreement->elements; // Illuminate\Database\Eloquent\Collection (not available in JS)

        $this->agreementCollection = Agreement::all(); // Illuminate\Database\Eloquent\Collection (not available in JS)
    }
}
```

**Context**
- Livewire version: 2.2.4
- Laravel version: 8.4.0
- Browser: All

**Comment by jlindsey3** [comment]
I have had trouble with collections as well. Always had to call `$collection->toArray()` and just figured I was doing something wrong and couldn't directly use a collection.

**Comment by StefanSchuechl** [comment]
I created a Playground for this issue:
https://laravelplayground.com#/snippets/cea334c6-dd67-4390-b294-b3c79e90cffd

The initial rendering is working. After the first roundtrip the Collection comes back as array.

**Comment by silose** [comment]
Got same problem, when passing collection with relations not updating

**Comment by basepack** [comment]
I got something similar:

```php
class ProductSelect extends Component
{
   
    public ?Collection $foundProducts;
}
```

Blade view:
```blade
@if($foundProducts->isEmpty())
       <div>No products found</div>
@endif
```

Gives the error: `Call to a member function isEmpty() on null` or `TypeError
Typed property App\Http\Livewire\Input\ProductSelect::$foundProducts must be an instance of Illuminate\Support\Collection, null used`

It is fixed if I supplement the component with casting:
```php
protected $casts = [
    'foundProducts' => 'collection',
];
```

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by basepack** [comment]
REOPEN

**Comment by KenLewis** [comment]
@calebporzio I haven't had time to learn how Livewire supposed to handle this since I've put it in. If you need more details, please let me know. I'd be more than happy to work with you on this.

**Comment by IP-Developer** [comment]
Got the same issue. Setting an Eloquent collection as a property results in empty array in JS, but it works as expected in Blade. So if I have a collection of 40 models, in JS I get an array with 40 empty arrays inside.

**Comment by joshhanley** [comment]
Eloquent Collection data won't be made available to the front end unless you have the collection properties specified in the `$rules` array as the contents are models, so the same conditions apply to it as apply to just a single model property. This is for security reasons so the whole collection and all properties aren't just sent to the front end.

See the example here of binding to an eloquent collection https://laravel-livewire.com/docs/2.x/properties#binding-models
> Note: For this to work, you have a validation entry in the `$rules` property for any model attributes you want to bind to. Otherwise, an error will be thrown.
> Additionally, you can bind to models within an Eloquent Collection.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by ronald2wing** [comment]
Hello, this issue still exists in Volt. The model is not available in $wire even if the rules are defined. I will use an array for now.

**Comment by joshhanley** [comment]
@ronald2wing yes thats correct, in V3 it has been disabled. We don't recommend using it going forward, but if you need to, to support upgrading an older app, you can enable `legacy_model_binding` in Livewire config.
